src/portable/st/synopsys/dcd_synopsys.c

```diff
@@ -592,29 +607,14 @@ bool dcd_edpt_open (uint8_t rhport, tusb_desc_endpoint_t const * desc_edpt)
     //    - Interrupt is EPSize
     //    - Bulk/ISO is max(EPSize, remaining-fifo / non-opened-EPIN)
```

Comment:
True, i removed it!
If a bigger FIFO results in a higher throughput, what is the matter to define a bigger EP size anyway? I mean, why should i set an EP size lower than the FIFO size? As for the STM32, you can not write anything into the FIFO once it was scheduled to transmit. Only after the FIFO is empty you can put more into it... But i am not an expert here - it is easy to manipulate the sizes by additional settings. I recommend doing this when somebody really has the need/time to benchmark it.

Reply:
@PanRe the EP Size is USB specs's transaction size, each packet can only carry that much. For Fullspeed you can only have max of 64 bytes, but you could set the FIFO for it to larger value e.g 256. The screenshot probably help to explain it, the MSC Read a very large data e.g 4096 bytes, since this is FS the EP size is = 64. 

- when FIFO size = 64, hw will stop receive data and NAK until software stack (tinyusb) handle that fifo level full interrupt, which is mostly occurred at the end of usb frame (1 ms) --> throughput is 64 byte per ms
- when FIFO size = 512, hw will keep receiving it  until FIFO is full and trigger interrupt --> throughput is 512 bytes per ms.

There is similar discussion on this throughput in this PR #399 with a very helpful screenshot from oscilloscope.

![image](https://user-images.githubusercontent.com/249515/103561984-f2f01200-4eec-11eb-8b8b-50f4be25ab03.png)

Reply:
however, you don't have to worry about this for now, I will try to come up with a scheme for these type of optimization later on. The most demanding throughput is MSC, it is normal for MSC to read/write 64KB or more in a single SCSI command.

Reply:
Very interesting! Thx for the details! Well maybe we should already optimize the FIFO allocation a bit now before complains about performance comes up.
Since we do not parse the descriptors, there is now way of determining the optimal FIFO size. IMHO we need additional information from the user how much bigger the FIFO compared to the EP size should be!

How about introducing a factor `ff_fact` which determines the FIFO size compares to the EP size. There are various possibilities e.g.
- Define factors for EP types - for ISO EPs ff_fact should always be 1 (bigger makes no sense), BULK should always be X etc.
- Define a factor array `uint8_t ff_fact[EP_MAX]` where the user can determine the factors explicitly
- Make a second usecase by a #define where the user can explicitly define the FIFO sizes in words

However, all of them may end up making problems when new configurations are set up where a completly different allocation optimization is necessary... But those are special cases where the user should be able to hack his requirements into the allocation scheme on his own.

What do you think of this?

Reply:
@PanRe don't worry about it for now, people haven't complained just yet. There is a few ways we could bypass the auto allocation like some specific macros as you mentioned, or having platform-dependent API, I didn't make up my mind just yet. 

> P.S.: Is it alway the case that BULK EPs work better the bigger their FIFO is? How about CTRL and INTERRUPT EPs? Do they also gain significantly from this? ISO does not since they are transmitted only once per frame.

Only Bulk and Control (asynchronous transfer) are able to carry more than 1 packet per frame. Interrupt and ISO (synchronous transfer)  has exactly 1 packet per frame. Control transfer is mostly used for enumeration and management, the payload is in general is very small 95% is less than 64 bytes except for DFU class where control is used to transfer data. So if we need to optimize, Bulk is where we should do.

PS: It is not urgent, and I think we should leave that as a separate PR where we could have an setup pre & post PR to test out the throughput.

Reply:
Ok fine! I just thought that if we would find a way to solve the fragmentation problem with the TX FIFOs (as for now only the last opened IN EP can be closed) then we could also stepwise optimize the FIFO allocation! To solve the fragmentation problem we would need to find a way to reallocate active IN EP FIFOs at runtime. This could also be used for optimization. For example, lets say CTRL, INT, and ISO EPs get only the FIFO size equal to the EP size, so no problems with opening them. If a BULK EP is opened it gets the total remaining space available. If after a BULK EP was opened an additional EP to be opened needs some space, we would need to reduce the size of the BULK EP and once done open the new EP. Of course a BULK EP needs a minimum size which at the end determines if an allocation overflow happens.

Relocating IN EPs, however, must be done consecutively, starting with the most upper located FIFO. The relocation can be done when the transmit FIFO empty interrupt fires, telling you its empty. In this case you can close the EP and reopen it at the new location. However, this means we may need to wait a couple of frames until the FIFO is empty. This on the other hand postpones the opening of an EP which waits until free space is available and i have no clue if the USB framework allows for such delays.

Just a thought of mine, if somebody has the time to sovle the relocation problem he also solves the optimal allocation problem.

Reply:
@PanRe those are really great idea, the open/close endpoint does indeed complicate the optimal fifo allocation for Bulk EP. Hopefully people don't complain that fast about the throughput on stm32 :)

Reply:
As a very quick workaround we could introduce a factor for BULK and CTRL EPs by use of a #define. This way, the users can help themself by adjusting them.

Reply:
That is one of the way to do it, however, currently the stack doesn't have any special macro for any dcd port since I want to abstract it out as much as possible.

Reply:
I had some thoughts about how to solve the IN EP relocation problem and before i can start with that i would need to be sure about one USB framework behaviour:
If a host tells a device to open an EP (IN or OUT does not matter) and the device currently is not able to execute this request it responds with a NAK message. Some time later, the host will try again and resends its request. Is this correct so far? And further, does the host in the meantime handle remaining active EPs? If so, IN EP FIFO relocations could be less complicated than thought. I know all of the above can be found within the USB specs but i would need hours verifying my questions and you may already know the answer to my questions. Thanks in advance!

Reply:
Not quite like that, the host doesn't explicitly tell device to open endpoints, all Control-Bulk-Interrupt (CBI) must be opened after the SET_CONFIGURE request. EP needs to be opened/enabled in order to response with NAK, For ISO, since default ITF is almost always has no endpoint. Host will need to SET_INTERFACE to switch to appropriate profile (endpoint size) before communicating.

In short, all CBI must be opened with SET_CONFIGURE, if device failed to response to this request, host may retry a couple of time and eventual give up as failed enumeration.